File: frontend/src/types/report.ts
export interface ReportMetadata {
  report_title: string
  dataset_name: string
  dataset_id: string
  generated_at: string
  domain_id: string
  domain_name: string
  row_count: number
  column_count: number
  reporting_period: string
  currency_symbol?: string | null
  data_quality_score: number
  available_sections_count: number
  total_sections_count: number
}

export interface ExecutiveSummary {
  dataset_name: string
  row_count: number
  column_count: number
  domain_name: string
  reporting_period: string
  currency_symbol?: string | null
  overall_performance_summary: string
  overall_business_condition?: string
  total_sales_revenue?: string | null
  total_cost_expense?: string | null
  total_profit_loss?: string | null
  profit_margin_pct?: string | null
  growth_or_decline_text?: string | null
  total_orders_count?: string | null
  average_order_value?: string | null
  total_units_sold?: string | null
  average_discount_pct?: string | null
  overall_business_health: string
  major_positive_findings?: string[]
  major_negative_findings?: string[]
  key_findings: string[]
  most_important_risks: string[]
  priority_corrective_actions: string[]
  top_3_recommended_actions?: string[]
}

export interface ColumnSummaryItem {
  column_name: string
  dtype: string
  semantic_role: string
  missing_count: number
  missing_pct: number
  unique_count: number
  sample_values: string[]
}

export interface DatasetOverviewReport {
  dimensions_text: string
  row_count: number
  column_count: number
  columns_summary: ColumnSummaryItem[]
  detected_date_range?: string | null
  has_date_dimension: boolean
  data_quality_score: number
  data_quality_status: "Healthy" | "Warning" | "Critical"
  missing_cells_count: number
  missing_cells_pct: number
  duplicate_rows_count: number
  duplicate_rows_pct: number
  important_limitations: string[]
}

export interface SixPointFinding {
  observation_title: string
  what_happened: string
  business_meaning: string
  why_it_matters: string
  recommended_action: string
  evidence: string
  limitations: string
}

export interface PerformancePerformerItem {
  entity_type: string
  name: string
  metric_name: string
  formatted_value: string
  raw_value?: number | null
  percentage_share?: number | null
  note?: string | null
}

export interface PerformanceMetricItem {
  name: string
  formatted_value: string
  raw_value?: number | null
  unit?: string | null
  status: "Available" | "Calculated" | "Estimated" | "Unavailable"
  business_meaning: string
  source_column?: string | null
  comparison_with_previous?: string | null
  requires_attention?: boolean
  attention_reason?: string | null
}

export interface BusinessPerformanceReport {
  is_available: boolean
  unavailable_reason?: string | null
  total_sales_revenue?: string | null
  total_profit_loss?: string | null
  profit_margin_pct?: string | null
  average_order_value?: string | null
  growth_rate_pct?: string | null
  performance_direction?: "growth" | "decline" | "stable" | "neutral" | null
  metric_highlights: PerformanceMetricItem[]
  top_performers: PerformancePerformerItem[]
  underperformers: PerformancePerformerItem[]
  six_point_findings?: SixPointFinding[]
  summary_text: string
}

export interface ProfitLossSegmentItem {
  name: string
  segment_type: string
  revenue_formatted: string
  profit_loss_formatted: string
  profit_margin_pct: number
  discount_rate_pct?: number | null
  issue_type: "loss_maker" | "high_sales_low_profit" | "discount_bleed" | "healthy"
  explanation: string
  calculation_basis?: string | null
  status_type?: "confirmed_finding" | "signal_requiring_validation"
  actionable_response?: string | null
}

export interface ProfitLossReport {
  is_available: boolean
  unavailable_reason?: string | null
  revenue_trend_summary: string
  profit_trend_summary: string
  profit_margin_summary: string
  loss_making_segments: ProfitLossSegmentItem[]
  high_sales_low_profit_segments: ProfitLossSegmentItem[]
  discount_margin_observations: string[]
  changes_explanation: string[]
}

export interface CustomerSegmentItem {
  segment_name: string
  customer_count_formatted?: string | null
  revenue_formatted: string
  revenue_share_pct: number
  aov_formatted: string
  profit_formatted?: string | null
  profit_margin_pct?: number | null
  business_implication: string
  marketing_strategy: string
}

export interface CustomerSegmentAnalysis {
  is_available: boolean
  unavailable_reason?: string | null
  dimension_analyzed?: string | null
  segments: CustomerSegmentItem[]
  concentration_observation?: string | null
  strategic_summary: string
}

export interface RegionalChannelItem {
  name: string
  dimension_type: string
  revenue_formatted: string
  revenue_share_pct: number
  order_volume: number
  operational_observation: string
}

export interface RegionalChannelAnalysis {
  is_available: boolean
  unavailable_reason?: string | null
  dimension_name?: string | null
  items: RegionalChannelItem[]
  operational_takeaway: string
}

export interface RootCauseItem {
  issue_title: string
  confirmed_observation: string
  possible_contributing_factors: string[]
  data_needed_for_validation: string[]
  recommended_investigation: string
}

export interface RootCauseAnalysisReport {
  is_available: boolean
  summary_statement: string
  root_causes: RootCauseItem[]
}

export interface DiscountPricingReport {
  is_available: boolean
  unavailable_reason?: string | null
  average_discount_pct?: string | null
  high_discount_revenue?: string | null
  high_discount_transaction_count: number
  margin_erosion_estimate?: string | null
  discount_tier_observations: string[]
  suggested_controls: string[]
  validation_requirements?: string[]
}

export interface InventoryReport {
  is_available: boolean
  unavailable_reason?: string | null
  total_stock_units?: string | null
  low_stock_count: number
  overstocked_count: number
  inventory_turnover_observation: string
  reorder_alerts: string[]
  inventory_items: Array<Record<string, any>>
}

export interface TrendPeriodComparisonItem {
  metric: string
  time_period: string
  previous_value?: number | null
  previous_value_formatted: string
  current_value?: number | null
  current_value_formatted: string
  absolute_change?: number | null
  absolute_change_formatted: string
  pct_change?: number | null
  pct_change_formatted: string
  is_positive: boolean
  direction: "increasing" | "decreasing" | "stable" | "fluctuating"
}

export interface TrendsReport {
  is_available: boolean
  unavailable_reason?: string | null
  analyzed_metric?: string | null
  time_column?: string | null
  time_period?: string | null
  selected_granularity: string
  period_comparison?: TrendPeriodComparisonItem | null
  upward_trends: string[]
  downward_trends: string[]
  peak_period?: string | null
  peak_value_formatted?: string | null
  lowest_period?: string | null
  lowest_value_formatted?: string | null
  seasonality_notes?: string | null
  volatility_assessment: string
  plain_language_interpretation: string
  limitations: string[]
}

export interface ForecastPointItem {
  period: string
  forecast_formatted: string
  forecast_raw: number
  lower_bound_80_formatted: string
  upper_bound_80_formatted: string
  lower_bound_95_formatted: string
  upper_bound_95_formatted: string
}

export interface ForecastReport {
  is_available: boolean
  unavailable_reason?: string | null
  forecasted_metric?: string | null
  time_horizon_periods: number
  frequency_label: string
  method_used: string
  latest_actual_formatted?: string | null
  final_forecast_formatted?: string | null
  projected_change_formatted?: string | null
  projected_growth_pct?: number | null
  forecast_points: ForecastPointItem[]
  confidence_score: number
  forecast_limitations: string[]
  plain_language_interpretation: string
  disclaimer: string
}

export interface VerifiedRiskItem {
  risk_id: string
  title: string
  severity: "low" | "medium" | "high" | "critical"
  category: string
  evidence: string
  affected_metric_or_segment: string
  business_impact: string
  why_it_matters?: string
  recommended_action: string
  qualification?: string | null
  data_required_for_validation?: string
  confidence_level?: string
}

export interface RiskDistributionInsight {
  dimension: string
  dominant_category: string
  percentage: number
  description: string
  observation_note: string
}

export interface RisksReport {
  is_available: boolean
  unavailable_reason?: string | null
  health_status: "Healthy" | "Attention Required" | "Critical Risks Identified"
  summary_statement: string
  total_risks_count: number
  high_severity_count: number
  medium_severity_count: number
  low_severity_count: number
  verified_risks: VerifiedRiskItem[]
  distribution_insights: RiskDistributionInsight[]
  data_safety_notes: string[]
}

export interface CompetitorComparisonItem {
  rank: number
  name: string
  is_our_entity: boolean
  revenue_formatted: string
  profit_formatted: string
  profit_margin_pct?: number | null
  market_share_pct?: number | null
  growth_rate_pct?: number | null
  pricing_index_formatted: string
  status_label: string
}

export interface CompetitionReport {
  is_available: boolean
  status_title: string
  summary_statement: string
  unavailable_reason?: string | null
  has_external_benchmark: boolean
  benchmark_filename?: string | null
  total_competitors_tracked: number
  competitors: CompetitorComparisonItem[]
  market_gaps: string[]
  strategic_actions: string[]
  required_market_fields: Array<{
    field: string
    type: string
    importance: string
    purpose: string
  }>
  market_limitations: string[]
}

export interface UnifiedRecommendationItem {
  rec_id: string
  title: string
  category: string
  priority: "critical" | "high" | "medium" | "low"
  problem: string
  evidence: string
  business_impact: string
  exact_recommended_action: string
  action_steps?: string[]
  expected_objective: string
  suggested_owner_team: string
  metric_to_track: string
  suggested_review_period: string
  data_limitations: string
}

export interface RecommendationsReport {
  is_available: boolean
  unavailable_reason?: string | null
  total_recommendations: number
  critical_count: number
  high_count: number
  medium_count: number
  low_count: number
  summary_statement: string
  recommendations_list: UnifiedRecommendationItem[]
}

export interface CorrectiveActionItem {
  priority: "Critical" | "High" | "Medium" | "Low"
  problem: string
  recommended_action: string
  owner_team: string
  metric_to_track: string
  review_period: string
}

export interface CorrectiveActionPlan {
  is_available: boolean
  action_items: CorrectiveActionItem[]
  disclaimer: string
}

export interface ConclusionReport {
  main_findings: string[]
  most_urgent_issues: string[]
  opportunities: string[]
  recommended_next_steps: string[]
  important_limitations: string[]
  areas_requiring_additional_data: string[]
}

export interface MethodologyAuditTrail {
  domain_rule_applied: string
  statistical_checks_performed: string[]
  granularity_grain_description: string
  evidence_baseline: string
  generated_timestamp: string
}

export interface ComprehensiveReport {
  metadata: ReportMetadata
  executive_summary: ExecutiveSummary
  business_performance: BusinessPerformanceReport
  profit_loss: ProfitLossReport
  customer_segment_analysis?: CustomerSegmentAnalysis | null
  regional_channel_analysis?: RegionalChannelAnalysis | null
  discount_pricing?: DiscountPricingReport | null
  inventory_analysis?: InventoryReport | null
  root_cause_analysis?: RootCauseAnalysisReport | null
  trends_intelligence: TrendsReport
  risks_anomalies: RisksReport
  market_competition: CompetitionReport
  recommendations: RecommendationsReport
  corrective_action_plan: CorrectiveActionPlan
  forecasting: ForecastReport
  data_driven_conclusion: ConclusionReport
  dataset_overview: DatasetOverviewReport
  methodology_audit?: MethodologyAuditTrail | null
}

export type ComprehensiveReportResponse = ComprehensiveReport

File: backend/app/schemas/report.py
"""Report Schemas for DataScope Unified Comprehensive Reporting Engine.

Defines Pydantic data structures for executive summaries, dataset overviews,
business performance, profit & loss, discount dynamics, inventory analysis,
trends, forecasting, risks, market competition, recommendations, corrective action plans,
conclusions, and document export metadata.
"""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import BaseModel, Field


class ReportMetadataSchema(BaseModel):
    report_title: str
    dataset_name: str
    dataset_id: str
    generated_at: str
    domain_id: str
    domain_name: str
    row_count: int = 0
    column_count: int = 0
    reporting_period: str
    currency_symbol: Optional[str] = None
    data_quality_score: int = 100
    available_sections_count: int = 0
    total_sections_count: int = 14


class ExecutiveSummarySchema(BaseModel):
    dataset_name: str
    row_count: int = 0
    column_count: int = 0
    domain_name: str
    reporting_period: str
    currency_symbol: Optional[str] = None
    overall_performance_summary: str
    overall_business_condition: str = "Healthy"
    total_sales_revenue: Optional[str] = None
    total_cost_expense: Optional[str] = None
    total_profit_loss: Optional[str] = None
    profit_margin_pct: Optional[str] = None
    growth_or_decline_text: Optional[str] = None
    total_orders_count: Optional[str] = None
    average_order_value: Optional[str] = None
    total_units_sold: Optional[str] = None
    average_discount_pct: Optional[str] = None
    overall_business_health: str = "Healthy"
    major_positive_findings: List[str] = Field(default_factory=list)
    major_negative_findings: List[str] = Field(default_factory=list)
    key_findings: List[str] = Field(default_factory=list)
    most_important_risks: List[str] = Field(default_factory=list)
    priority_corrective_actions: List[str] = Field(default_factory=list)
    top_3_recommended_actions: List[str] = Field(default_factory=list)


class ColumnSummaryItemSchema(BaseModel):
    column_name: str
    dtype: str
    semantic_role: str = "general"
    missing_count: int = 0
    missing_pct: float = 0.0
    unique_count: int = 0
    sample_values: List[str] = Field(default_factory=list)


class DatasetOverviewReportSchema(BaseModel):
    dimensions_text: str = ""
    row_count: int = 0
    column_count: int = 0
    columns_summary: List[ColumnSummaryItemSchema] = Field(default_factory=list)
    detected_date_range: Optional[str] = None
    has_date_dimension: bool = False
    data_quality_score: int = 100
    data_quality_status: Literal["Healthy", "Warning", "Critical"] = "Healthy"
    missing_cells_count: int = 0
    missing_cells_pct: float = 0.0
    duplicate_rows_count: int = 0
    duplicate_rows_pct: float = 0.0
    important_limitations: List[str] = Field(default_factory=list)


class SixPointFindingSchema(BaseModel):
    observation_title: str
    what_happened: str
    business_meaning: str
    why_it_matters: str
    recommended_action: str
    evidence: str
    limitations: str


class PerformancePerformerItemSchema(BaseModel):
    entity_type: str  # e.g., "Category", "Product", "Region", "Team", "Segment"
    name: str
    metric_name: str
    formatted_value: str
    raw_value: Optional[float] = None
    percentage_share: Optional[float] = None
    note: Optional[str] = None


class PerformanceMetricItemSchema(BaseModel):
    name: str
    formatted_value: str
    raw_value: Optional[float] = None
    unit: Optional[str] = None
    status: Literal["Available", "Calculated", "Estimated", "Unavailable"] = "Available"
    business_meaning: str = ""
    source_column: Optional[str] = None
    comparison_with_previous: Optional[str] = None
    requires_attention: bool = False
    attention_reason: Optional[str] = None


class BusinessPerformanceReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    total_sales_revenue: Optional[str] = None
    total_profit_loss: Optional[str] = None
    profit_margin_pct: Optional[str] = None
    average_order_value: Optional[str] = None
    growth_rate_pct: Optional[str] = None
    performance_direction: Optional[Literal["growth", "decline", "stable", "neutral"]] = None
    metric_highlights: List[PerformanceMetricItemSchema] = Field(default_factory=list)
    top_performers: List[PerformancePerformerItemSchema] = Field(default_factory=list)
    underperformers: List[PerformancePerformerItemSchema] = Field(default_factory=list)
    six_point_findings: List[SixPointFindingSchema] = Field(default_factory=list)
    summary_text: str = ""


class ProfitLossSegmentItemSchema(BaseModel):
    name: str
    segment_type: str  # "Product", "Category", "Region", "Sub-Category"
    revenue_formatted: str
    profit_loss_formatted: str
    profit_margin_pct: float
    discount_rate_pct: Optional[float] = None
    issue_type: Literal["loss_maker", "high_sales_low_profit", "discount_bleed", "healthy"] = "loss_maker"
    explanation: str
    calculation_basis: Optional[str] = None
    status_type: Literal["confirmed_finding", "signal_requiring_validation"] = "confirmed_finding"
    actionable_response: Optional[str] = None


class ProfitLossReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    revenue_trend_summary: str = ""
    profit_trend_summary: str = ""
    profit_margin_summary: str = ""
    loss_making_segments: List[ProfitLossSegmentItemSchema] = Field(default_factory=list)
    high_sales_low_profit_segments: List[ProfitLossSegmentItemSchema] = Field(default_factory=list)
    discount_margin_observations: List[str] = Field(default_factory=list)
    changes_explanation: List[str] = Field(default_factory=list)


class CustomerSegmentItemSchema(BaseModel):
    segment_name: str
    customer_count_formatted: Optional[str] = None
    revenue_formatted: str
    revenue_share_pct: float
    aov_formatted: str
    profit_formatted: Optional[str] = None
    profit_margin_pct: Optional[float] = None
    business_implication: str
    marketing_strategy: str


class CustomerSegmentAnalysisSchema(BaseModel):
    is_available: bool = False
    unavailable_reason: Optional[str] = None
    dimension_analyzed: Optional[str] = None
    segments: List[CustomerSegmentItemSchema] = Field(default_factory=list)
    concentration_observation: Optional[str] = None
    strategic_summary: str = ""


class RegionalChannelItemSchema(BaseModel):
    name: str
    dimension_type: str  # "Region", "Channel", "Device"
    revenue_formatted: str
    revenue_share_pct: float
    order_volume: int
    operational_observation: str


class RegionalChannelAnalysisSchema(BaseModel):
    is_available: bool = False
    unavailable_reason: Optional[str] = None
    dimension_name: Optional[str] = None
    items: List[RegionalChannelItemSchema] = Field(default_factory=list)
    operational_takeaway: str = ""


class RootCauseItemSchema(BaseModel):
    issue_title: str
    confirmed_observation: str
    possible_contributing_factors: List[str] = Field(default_factory=list)
    data_needed_for_validation: List[str] = Field(default_factory=list)
    recommended_investigation: str


class RootCauseAnalysisReportSchema(BaseModel):
    is_available: bool = True
    summary_statement: str = ""
    root_causes: List[RootCauseItemSchema] = Field(default_factory=list)


class DiscountPricingReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    average_discount_pct: Optional[str] = None
    high_discount_revenue: Optional[str] = None
    high_discount_transaction_count: int = 0
    margin_erosion_estimate: Optional[str] = None
    discount_tier_observations: List[str] = Field(default_factory=list)
    suggested_controls: List[str] = Field(default_factory=list)
    validation_requirements: List[str] = Field(default_factory=list)


class InventoryReportSchema(BaseModel):
    is_available: bool = False
    unavailable_reason: Optional[str] = None
    total_stock_units: Optional[str] = None
    low_stock_count: int = 0
    overstocked_count: int = 0
    inventory_turnover_observation: str = ""
    reorder_alerts: List[str] = Field(default_factory=list)
    inventory_items: List[Dict[str, Any]] = Field(default_factory=list)


class TrendPeriodComparisonItemSchema(BaseModel):
    metric: str
    time_period: str
    previous_value: Optional[float] = None
    previous_value_formatted: str = "N/A"
    current_value: Optional[float] = None
    current_value_formatted: str = "N/A"
    absolute_change: Optional[float] = None
    absolute_change_formatted: str = "N/A"
    pct_change: Optional[float] = None
    pct_change_formatted: str = "N/A"
    is_positive: bool = True
    direction: Literal["increasing", "decreasing", "stable", "fluctuating"] = "stable"


class TrendsReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    analyzed_metric: Optional[str] = None
    time_column: Optional[str] = None
    time_period: Optional[str] = None
    selected_granularity: str = "auto"
    period_comparison: Optional[TrendPeriodComparisonItemSchema] = None
    upward_trends: List[str] = Field(default_factory=list)
    downward_trends: List[str] = Field(default_factory=list)
    peak_period: Optional[str] = None
    peak_value_formatted: Optional[str] = None
    lowest_period: Optional[str] = None
    lowest_value_formatted: Optional[str] = None
    seasonality_notes: Optional[str] = None
    volatility_assessment: str = "Stable"
    plain_language_interpretation: str = ""
    limitations: List[str] = Field(default_factory=list)


class ForecastPointItemSchema(BaseModel):
    period: str
    forecast_formatted: str
    forecast_raw: float
    lower_bound_80_formatted: str
    upper_bound_80_formatted: str
    lower_bound_95_formatted: str
    upper_bound_95_formatted: str


class ForecastReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    forecasted_metric: Optional[str] = None
    time_horizon_periods: int = 7
    frequency_label: str = "Daily"
    method_used: str = "Statistical Holt's Smoothing"
    latest_actual_formatted: Optional[str] = None
    final_forecast_formatted: Optional[str] = None
    projected_change_formatted: Optional[str] = None
    projected_growth_pct: Optional[float] = None
    forecast_points: List[ForecastPointItemSchema] = Field(default_factory=list)
    confidence_score: float = 0.85
    forecast_limitations: List[str] = Field(default_factory=list)
    plain_language_interpretation: str = ""
    disclaimer: str = (
        "Statistical forecasts extrapolate historical patterns without guaranteeing future outcomes or modeling unpredictable external factors."
    )


class VerifiedRiskItemSchema(BaseModel):
    risk_id: str
    title: str
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    category: str
    evidence: str
    affected_metric_or_segment: str
    business_impact: str
    why_it_matters: str = ""
    recommended_action: str
    qualification: Optional[str] = None
    data_required_for_validation: str = ""
    confidence_level: str = "High (Verified from Dataset Evidence)"


class RiskDistributionInsightSchema(BaseModel):
    dimension: str
    dominant_category: str
    percentage: float
    description: str
    observation_note: str


class RisksReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    health_status: Literal["Healthy", "Attention Required", "Critical Risks Identified"] = "Healthy"
    summary_statement: str = ""
    total_risks_count: int = 0
    high_severity_count: int = 0
    medium_severity_count: int = 0
    low_severity_count: int = 0
    verified_risks: List[VerifiedRiskItemSchema] = Field(default_factory=list)
    distribution_insights: List[RiskDistributionInsightSchema] = Field(default_factory=list)
    data_safety_notes: List[str] = Field(default_factory=list)


class CompetitorComparisonItemSchema(BaseModel):
    rank: int
    name: str
    is_our_entity: bool = False
    revenue_formatted: str = "N/A"
    profit_formatted: str = "N/A"
    profit_margin_pct: Optional[float] = None
    market_share_pct: Optional[float] = None
    growth_rate_pct: Optional[float] = None
    pricing_index_formatted: str = "N/A"
    status_label: str = "Reported Competitor"


class CompetitionReportSchema(BaseModel):
    is_available: bool = False
    status_title: str = "Market competition analysis is not available."
    summary_statement: str = "No external market competitor data was provided."
    unavailable_reason: Optional[str] = None
    has_external_benchmark: bool = False
    benchmark_filename: Optional[str] = None
    total_competitors_tracked: int = 0
    competitors: List[CompetitorComparisonItemSchema] = Field(default_factory=list)
    market_gaps: List[str] = Field(default_factory=list)
    strategic_actions: List[str] = Field(default_factory=list)
    required_market_fields: List[Dict[str, Any]] = Field(default_factory=list)
    market_limitations: List[str] = Field(default_factory=list)


class UnifiedRecommendationItemSchema(BaseModel):
    rec_id: str
    title: str
    category: str
    priority: Literal["critical", "high", "medium", "low"] = "medium"
    problem: str
    evidence: str
    business_impact: str
    exact_recommended_action: str
    action_steps: List[str] = Field(default_factory=list)
    expected_objective: str
    suggested_owner_team: str = "Operations"
    metric_to_track: str = "General Performance"
    suggested_review_period: str = "Monthly"
    data_limitations: str = ""


class RecommendationsReportSchema(BaseModel):
    is_available: bool = True
    unavailable_reason: Optional[str] = None
    total_recommendations: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    summary_statement: str = ""
    recommendations_list: List[UnifiedRecommendationItemSchema] = Field(default_factory=list)


class CorrectiveActionItemSchema(BaseModel):
    priority: Literal["Critical", "High", "Medium", "Low"] = "Medium"
    problem: str
    recommended_action: str
    owner_team: str  # e.g., "Marketing", "Finance", "Operations", "Sales", "Engineering"
    metric_to_track: str
    review_period: str  # e.g., "Immediate / 7 Days", "30 Days", "Quarterly"


class CorrectiveActionPlanSchema(BaseModel):
    is_available: bool = True
    action_items: List[CorrectiveActionItemSchema] = Field(default_factory=list)
    disclaimer: str = (
        "Responsibility areas are recommended strategic operational divisions rather than individual employee assignments."
    )


class ConclusionReportSchema(BaseModel):
    main_findings: List[str] = Field(default_factory=list)
    most_urgent_issues: List[str] = Field(default_factory=list)
    opportunities: List[str] = Field(default_factory=list)
    recommended_next_steps: List[str] = Field(default_factory=list)
    important_limitations: List[str] = Field(default_factory=list)
    areas_requiring_additional_data: List[str] = Field(default_factory=list)


class MethodologyAuditTrailSchema(BaseModel):
    domain_rule_applied: str
    statistical_checks_performed: List[str] = Field(default_factory=list)
    granularity_grain_description: str
    evidence_baseline: str
    generated_timestamp: str


class ComprehensiveReportResponse(BaseModel):
    metadata: ReportMetadataSchema
    executive_summary: ExecutiveSummarySchema
    business_performance: BusinessPerformanceReportSchema
    profit_loss: ProfitLossReportSchema
    customer_segment_analysis: Optional[CustomerSegmentAnalysisSchema] = None
    regional_channel_analysis: Optional[RegionalChannelAnalysisSchema] = None
    discount_pricing: Optional[DiscountPricingReportSchema] = None
    inventory_analysis: Optional[InventoryReportSchema] = None
    root_cause_analysis: Optional[RootCauseAnalysisReportSchema] = None
    trends_intelligence: TrendsReportSchema
    risks_anomalies: RisksReportSchema
    market_competition: CompetitionReportSchema
    recommendations: RecommendationsReportSchema
    corrective_action_plan: CorrectiveActionPlanSchema
    forecasting: ForecastReportSchema
    data_driven_conclusion: ConclusionReportSchema
    dataset_overview: DatasetOverviewReportSchema
    methodology_audit: Optional[MethodologyAuditTrailSchema] = None
